rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! A body-wrapping type that ensures data is being streamed faster than some lower limit.
+//!
+//! If data is being streamed too slowly, this body type will emit an error next time it's polled.
+
+mod throughput;
+
+use aws_smithy_async::rt::sleep::{AsyncSleep, SharedAsyncSleep};
+use aws_smithy_async::time::{SharedTimeSource, TimeSource};
+use aws_smithy_runtime_api::shared::IntoShared;
+use bytes::Buf;
+use http::HeaderMap;
+use std::collections::VecDeque;
+use std::fmt;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, SystemTime};
+use throughput::Throughput;
+
+// Chosen arbitrarily.
+const LOG_WINDOW_SIZE: usize = 16;
+
+pin_project_lite::pin_project! {
+    /// A body-wrapper that will ensure that the wrapped body is emitting bytes faster than some
+    /// `minimum_throughput`.
+    pub struct MinimumThroughputBody<InnerBody> {
+        #[pin]
+        inner: InnerBody,
+        // A record of when and how much data was read
+        throughput_logs: VecDeque<(SystemTime, u64)>,
```

Comment:
How large does this queue get in practice?

Reply:
On my machine, a 100MB download with a tick rate of 1s produces 6700 logs. The download completes in only a few secs.